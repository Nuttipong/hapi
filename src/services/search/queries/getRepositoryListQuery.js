import { Paging, PageButton } from "../../../common/paging";

export class IGetRepositoryListQuery {
  constructor() {
    if (this.constructor.name === "IGetRepositoryListQuery") {
      throw new Error("IGetRepositoryListQuery is abstract");
    }
  }

  execute(dto) {}
}

export class GetRepositoryListQuery extends IGetRepositoryListQuery {
  constructor(logger, githubApi) {
    super();
    if (this.constructor["instance"]) {
      return this.constructor["instance"];
    }

    this.logger = logger;
    this.githubApi = githubApi;
    this.searchUrl = new URL("http://localhost:3000/search");
    this.constructor["instance"] = this;
  }

  async execute(dto) {
    let results = {};
    try {
      const query = new Map();
      for (const key in dto) {
        query.set(key, dto[key]);
      }
      const headers = {
        "Content-Type": "application/json;charset=utf-8",
      };

      const resp = await this.githubApi.get(query, headers);
      // Github support only 1000 items
      const totalCount = resp.total_count > 1000 ? 1000 : resp.total_count;
      if (resp && resp.items) {
        results.items = resp.items;
        results.total_count = totalCount;
        results.pagination = this.setPaging(
          totalCount,
          parseInt(dto.page),
          parseInt(dto.per_page),
          parseInt(dto.range),
          dto.q
        );
      }
    } catch (err) {
      this.logger.error(`Fetch api error : ${err}`);
    } finally {
      return results;
    }
  }

  setPaging(totalCount, page, per_page, range, q) {
    const paging = new Paging(totalCount, page, per_page, range);
    return {
      ranges: paging.setRange().setUrl(PageButton.RANGE, (p) => {
        return this.buildUrlWithQuery(this.searchUrl, p, per_page, q, range);
      }).rangeBtn,
      next: paging.setNext().setUrl(PageButton.NEXT, (p) => {
        return this.buildUrlWithQuery(this.searchUrl, p, per_page, q, range);
      }).nextBtn,
      prev: paging.setPrev().setUrl(PageButton.PREV, (p) => {
        return this.buildUrlWithQuery(this.searchUrl, p, per_page, q, range);
      }).prevBtn,
      first: paging.setFirst().setUrl(PageButton.FISRT, (p) => {
        return this.buildUrlWithQuery(this.searchUrl, p, per_page, q, range);
      }).firstBtn,
      last: paging.setLast().setUrl(PageButton.LAST, (p) => {
        return this.buildUrlWithQuery(this.searchUrl, p, per_page, q, range);
      }).lastBtn,
    };
  }

  buildUrlWithQuery(searchUrl, page, per_page, q, range) {
    const params = new URLSearchParams(searchUrl.search);
    params.append("q", q);
    params.append("page", page);
    params.append("per_page", per_page);
    params.append("range", range);
    return `${searchUrl.href}?${params.toString()}`;
  }
}
