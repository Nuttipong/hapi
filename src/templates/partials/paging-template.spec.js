import Handlebars from "../helpers/index";
import HTML from "./paging.html";
import { GetRepositoryListQuery } from "../../services/search/queries/getRepositoryListQuery";

describe("Paging partial", () => {
  it("should always render with pagination model", () => {
    // given
    const tuple = [
      [100, 1, 10, 10, "nodejs"],
      [100, 11, 10, 10, "nodejs"],
    ];
    const svc = new GetRepositoryListQuery();
    const template = Handlebars.compile(HTML);

    for (const row of tuple) {
      const [total, page, limit, range, q] = row;
      const model = svc.setPaging(total, page, limit, range, q);
      const vm = {
        pagination: model,
        page,
        total_count: total,
        items: [],
      };

      // when
      const actual = template({ vm });

      // then
      expect(actual).toBeDefined();
      // and
      expect(actual).toMatchSnapshot();
    }
  });
});
