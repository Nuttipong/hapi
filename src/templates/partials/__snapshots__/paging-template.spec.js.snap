// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paging partial should always render with pagination model 1`] = `
"<div class=\\"pagination\\">
      <a href=\\"#\\" class=\\"active\\">1</a>
   <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;2&amp;per_page&#x3D;10&amp;range&#x3D;10\\">2</a>
   <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;3&amp;per_page&#x3D;10&amp;range&#x3D;10\\">3</a>
   <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;4&amp;per_page&#x3D;10&amp;range&#x3D;10\\">4</a>
   <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;5&amp;per_page&#x3D;10&amp;range&#x3D;10\\">5</a>
   <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;6&amp;per_page&#x3D;10&amp;range&#x3D;10\\">6</a>
   <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;7&amp;per_page&#x3D;10&amp;range&#x3D;10\\">7</a>
   <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;8&amp;per_page&#x3D;10&amp;range&#x3D;10\\">8</a>
   <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;9&amp;per_page&#x3D;10&amp;range&#x3D;10\\">9</a>
   <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;10&amp;per_page&#x3D;10&amp;range&#x3D;10\\">10</a>
   
  <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;2&amp;per_page&#x3D;10&amp;range&#x3D;10\\">&gt;</a>
   
  <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;10&amp;per_page&#x3D;10&amp;range&#x3D;10\\">&raquo;</a>
   
</div>
"
`;

exports[`Paging partial should always render with pagination model 2`] = `
"<div class=\\"pagination\\">
   
  <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;1&amp;per_page&#x3D;10&amp;range&#x3D;10\\">&laquo;</a>
   
  <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;10&amp;per_page&#x3D;10&amp;range&#x3D;10\\">&lt;</a>
    <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;2&amp;per_page&#x3D;10&amp;range&#x3D;10\\">2</a>
   <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;3&amp;per_page&#x3D;10&amp;range&#x3D;10\\">3</a>
   <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;4&amp;per_page&#x3D;10&amp;range&#x3D;10\\">4</a>
   <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;5&amp;per_page&#x3D;10&amp;range&#x3D;10\\">5</a>
   <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;6&amp;per_page&#x3D;10&amp;range&#x3D;10\\">6</a>
   <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;7&amp;per_page&#x3D;10&amp;range&#x3D;10\\">7</a>
   <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;8&amp;per_page&#x3D;10&amp;range&#x3D;10\\">8</a>
   <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;9&amp;per_page&#x3D;10&amp;range&#x3D;10\\">9</a>
   <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;10&amp;per_page&#x3D;10&amp;range&#x3D;10\\">10</a>
    
  <a href=\\"http://localhost:3000/search?q&#x3D;nodejs&amp;page&#x3D;10&amp;per_page&#x3D;10&amp;range&#x3D;10\\">&raquo;</a>
   
</div>
"
`;
